airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DetectStreamToFlush.java

```diff
@@ -0,0 +1,222 @@
+    // when we are closing or queues are very full, flush regardless of how few items are in the queue.
+    return isClosing.get() || isBuffer90Full ? 0 : QUEUE_FLUSH_THRESHOLD_BYTES;
+  }
+
+  // todo (cgardens) - improve prioritization by getting a better estimate of how much data running
+  // workers will process. we have access to their batch sizes after all!
+  /**
+   * Iterates over streams until it finds one that is ready to flush. Streams are ordered by priority.
+   * Return an empty optional if no streams are ready.
+   * <p>
+   * A stream is ready to flush if it either meets a size threshold or a time threshold. See
+   * {@link #isSizeTriggered(StreamDescriptor, long)} and {@link #isTimeTriggered(StreamDescriptor)}
+   * for details on these triggers.
+   *
+   * @param queueSizeThresholdBytes - the size threshold to use for determining if a stream is ready
+   *        to flush.
+   * @return the next stream to flush. if no stream is ready to flush, empty.
+   */
+  @VisibleForTesting
+  Optional<StreamDescriptor> getNextStreamToFlush(final long queueSizeThresholdBytes) {
+    for (final StreamDescriptor stream : orderStreamsByPriority(bufferDequeue.getBufferedStreams())) {
+      final ImmutablePair<Boolean, String> isTimeTriggeredResult = isTimeTriggered(stream);
+      final ImmutablePair<Boolean, String> isSizeTriggeredResult = isSizeTriggered(stream, queueSizeThresholdBytes);
+
+      final String debugString = String.format(
+          "trigger info: %s - %s, %s , %s",
+          stream.getNamespace(),
+          stream.getName(),
+          isTimeTriggeredResult.getRight(),
+          isSizeTriggeredResult.getRight());
+      log.debug("computed: {}", debugString);
+
+      if (isSizeTriggeredResult.getLeft() || isTimeTriggeredResult.getLeft()) {
+        log.info("flushing: {}", debugString);
+
+        return Optional.of(stream);
+      }
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * The time trigger is based on the last time a record was added to the queue. We don't want records
+   * to sit forever, even if the queue is not that full (bad for time to value for users). Also, the
+   * more time passes since a record was added, the less likely another record is coming (caveat is
+   * CDC where it's random).
+   * <p>
+   * This method also returns debug string with info that about the computation. We do it this way, so
+   * that the debug info that is printed is exactly what is used in the computation.
+   *
+   * @param stream stream
+   * @return is time triggered and a debug string
+   */
+  @VisibleForTesting
+  ImmutablePair<Boolean, String> isTimeTriggered(final StreamDescriptor stream) {
+    final Boolean isTimeTriggered = bufferDequeue.getTimeOfLastRecord(stream)
+        .map(time -> time.isBefore(Instant.now().minus(MAX_TIME_BETWEEN_REC_MIN, ChronoUnit.MINUTES)))
+        .orElse(false);
+
+    final String debugString = String.format("time trigger: %s", isTimeTriggered);
```

Comment:
How useful is this for debugging? I can only see this being useful since `getNextStreamToFlush` doesn't differentiate between `isSizedTriggered` or `isTimeTriggered` to log the universal `debugString` and you want to know if it was time or size triggered

There would be greater value with the `Instant.now()` and `bufferDequeue.getTimeOfLastRecord(stream).get()` values imo

Reply:
fair point. in practice, this is a simple trigger and we haven't seen a bug here yet. i'm sure we will one day and then want to log more, but i'm inclined to wait until then. the tradeoff is how huge the log lines gets so if we go with what you're suggesting you really need `now: 2023/05/30 12:00.000+06 record added at: 2023/05/30 11:00.000+06`.